File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

from collections import deque
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root: return 0
        
        self.large = float('-inf')
        
        def dfs(node, num):
            if not node: return
            
            if not node.left and not node.right:
                self.large = max(self.large, num)
            
            dfs(node.left, num+1)
            dfs(node.right, num+1)
        
        dfs(root, 1)
        
        return self.large

File: 104-maximum-depth-of-binary-tree/104-maximum-depth-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
    
    if(!root){
        return 0
    }
    
    let stk = [root]
    let lev = 0
    let max = 1
    
    while(stk.length>0){
        curr = stk.shift()
        max--
        if(curr.left){
            stk.push(curr.left)
        }
        if(curr.right){
            stk.push(curr.right)
        }
        if(max==0){
            lev++
            max = stk.length
        }        
    }
    
    return lev
};